### PR TITLE
feat: Add new ptn modules `azd/ml-hub-dependencies`.

### LINKED ISSUE #1227: [Module Proposal]: `avm/ptn/azd/ml-hub-dependencies`
### Check for previous/existing GitHub issues/module proposals

- [X] I have checked for previous/existing GitHub issues/module proposals.

### Check this module doesn't already exist in the module indexes

- [X] I have checked for that this module doesn't already exist in the module indexes.

### Bicep or Terraform?

Bicep

### Module Classification?

Pattern Module

### Module Name

avm/ptn/azd/ml-hub-dependencies

### Module Details

Please add a Pattern Module named `avm/ptn/azd/ml-hub-dependencies`, this module will create the following seven resources:

1. Key Vault.
2. Storage Account.
3. Log Analytics Workspace.
4. Application Insights instance.
5. Azure Container Registry (ACR).
6. Cognitive Service.
7. Search Service.

For details, please refer to azd: 
https://github.com/Azure/azure-dev/blob/main/templates/common/infra/bicep/core/ai/hub-dependencies.bicep

CC: @jongio

### Do you want to be the owner of this module?

No

### Module Owner's GitHub Username (handle)

jongio

### (Optional) Secondary Module Owner's GitHub Username (handle)

_No response_

### PR DESCRIPTION
**Pull request by Menghua1**
## Description
Fixes https://github.com/Azure/Azure-Verified-Modules/issues/1227

## Pipeline Reference

| Pipeline |
| -------- |
|  [avm.ptn.azd.ml-hub-dependencies](https://github.com/Menghua1/bicep-registry-modules/actions/workflows/avm.ptn.azd.ml-hub-dependencies.yml) |

## Type of Change

- [ ] Update to CI Environment or utilities (Non-module affecting changes)
- [x] Azure Verified Module updates:
  - [ ] Bugfix containing backwards-compatible bug fixes, and I have NOT bumped the MAJOR or MINOR version in `version.json`:
    - [ ] Someone has opened a bug report issue, and I have included "Closes #{bug_report_issue_number}" in the PR description.
    - [ ] The bug was found by the module author, and no one has opened an issue to report it yet.
  - [ ] Feature update backwards compatible feature updates, and I have bumped the MINOR version in `version.json`.
  - [ ] Breaking changes and I have bumped the MAJOR version in `version.json`.
  - [ ] Update to documentation

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [x] I have run `Set-AVMModule` locally to generate the supporting module files.
- [x] My corresponding pipelines / checks run clean and green without any errors or warnings

@jongio for notification.

### CODE REVIEW

**Comment by ReneHezser** [comment]
@Menghua1 
We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.

Please reach out if any support is needed.

**Comment by Menghua1** [comment]
> @Menghua1 We have added examples for Bicep parameter files to the Readme. This has been applied to all published modules but needs to be done for PRs as well. Can you please update your branch and run the Set-AVMModule utility as detailed [here](https://azure.github.io/Azure-Verified-Modules/contributing/bicep/bicep-contribution-flow/generate-bicep-module-files/). It is required for the validation pipeline to succeed and the contribution to be published.
> 
> Please reach out if any support is needed.

@ReneHezser Thank you for your guidance. I have updated branch and re-ran the Set-AVMModule utility.

File: .github/CODEOWNERS [hunk 1]
```diff
@@ -15,6 +15,7 @@
 /avm/ptn/azd/apim-api/ @Azure/avm-ptn-azd-apimapi-module-owners-bicep @Azure/avm-module-reviewers-bicep
 /avm/ptn/azd/container-apps/ @Azure/avm-ptn-azd-containerapps-module-owners-bicep @Azure/avm-module-reviewers-bicep
 /avm/ptn/azd/insights-dashboard/ @Azure/avm-ptn-azd-insightsdashboard-module-owners-bicep @Azure/avm-module-reviewers-bicep
+/avm/ptn/azd/ml-hub-dependencies/ @Azure/avm-ptn-azd-mlhubdependencies-module-owners-bicep @Azure/avm-module-reviewers-bicep
```
Comment (AlexanderSehr) [inline .github/CODEOWNERS:18]: Team is either not existing or is not in the correct hierachy

File: avm/ptn/azd/ml-hub-dependencies/main.bicep [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+        telemetry: {
+          type: 'String'
+          value: 'For more information, see https://aka.ms/avm/TelemetryInfo'
+        }
+      }
+    }
+  }
+}
+
+module keyVault 'br/public:avm/res/key-vault/vault:0.7.1' = {
```
Comment (AlexanderSehr) [inline avm/ptn/azd/ml-hub-dependencies/main.bicep:295]: Note that all references to AVM modules inside your module should pass the `enableTelemetry` value through

File: avm/ptn/azd/ml-hub-dependencies/main.bicep [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+    enableVaultForDeployment: enableVaultForDeployment
+    enableVaultForTemplateDeployment: enableVaultForTemplateDeployment
+    enablePurgeProtection: enablePurgeProtection
+    sku: keyVaultSku
+  }
+}
+
+module storageAccount 'br/public:avm/res/storage/storage-account:0.9.1' = {
+  name: '${uniqueString(deployment().name, location)}-storage'
+  params: {
```
Comment (AlexanderSehr) [inline avm/ptn/azd/ml-hub-dependencies/main.bicep:311]: Misses passthru of `enableTelemetry` parameter

File: avm/ptn/azd/ml-hub-dependencies/main.bicep [hunk 4]
```diff
@@ -0,0 +1,475 @@
...
+    blobServices: blobServices
+    fileServices: fileServices
+    queueServices: queueServices
+    tableServices: tableServices
+  }
+}
+
+module cognitiveServices 'br/public:avm/res/cognitive-services/account:0.7.0' = {
+  name: '${uniqueString(deployment().name, location)}-cognitive'
+  params: {
```
Comment (AlexanderSehr) [inline avm/ptn/azd/ml-hub-dependencies/main.bicep:328]: Misses passthru of `enableTelemetry` parameter

File: avm/ptn/azd/ml-hub-dependencies/main.bicep [hunk 5]
```diff
@@ -0,0 +1,475 @@
...
+    networkAcls: cognitiveServicesNetworkAcls
+    disableLocalAuth: cognitiveServicesDisableLocalAuth
+    sku: cognitiveServicesSku
+    deployments: cognitiveServicesDeployments
+  }
+}
+
+module logAnalytics 'br/public:avm/res/operational-insights/workspace:0.6.0' = if (!empty(logAnalyticsName)) {
+  name: '${uniqueString(deployment().name, location)}-loganalytics'
+  params: {
```
Comment (AlexanderSehr) [inline avm/ptn/azd/ml-hub-dependencies/main.bicep:344]: Misses passthru of `enableTelemetry` parameter

File: avm/ptn/azd/ml-hub-dependencies/main.bicep [hunk 6]
```diff
@@ -0,0 +1,475 @@
...
+    location: location
+    tags: tags
+    dataRetention: dataRetention
+    skuName: logAnalyticsSkuName
+  }
+}
+
+module applicationInsights 'br/public:avm/ptn/azd/insights-dashboard:0.1.0' = if (!empty(applicationInsightsName) && !empty(logAnalyticsName)) {
+  name: '${uniqueString(deployment().name, location)}-insights'
+  params: {
```
Comment (AlexanderSehr) [inline avm/ptn/azd/ml-hub-dependencies/main.bicep:355]: Misses passthru of `enableTelemetry` parameter

File: avm/ptn/azd/ml-hub-dependencies/main.bicep [hunk 7]
```diff
@@ -0,0 +1,475 @@
...
+    tags: tags
+    name: applicationInsightsName
+    dashboardName: applicationInsightsDashboardName
+    logAnalyticsWorkspaceResourceId: !empty(logAnalyticsName) ? logAnalytics.outputs.resourceId : ''
+  }
+}
+
+module containerRegistry 'br/public:avm/res/container-registry/registry:0.4.0' = if (!empty(containerRegistryName)) {
+  name: '${uniqueString(deployment().name, location)}-registry'
+  params: {
```
Comment (AlexanderSehr) [inline avm/ptn/azd/ml-hub-dependencies/main.bicep:366]: Misses passthru of `enableTelemetry` parameter

File: avm/ptn/azd/ml-hub-dependencies/main.bicep [hunk 8]
```diff
@@ -0,0 +1,475 @@
...
+    acrSku: registryAcrSku
+    tags: tags
+    location: location
+    publicNetworkAccess: registryPublicNetworkAccess
+  }
+}
+
+module searchService 'br/public:avm/res/search/search-service:0.6.0' = if (!empty(searchServiceName)) {
+  name: '${uniqueString(deployment().name, location)}-searchservice'
+  params: {
```
Comment (AlexanderSehr) [inline avm/ptn/azd/ml-hub-dependencies/main.bicep:377]: Misses passthru of `enableTelemetry` parameter

File: .github/CODEOWNERS (hunk 1 above)
Comment (AlexanderSehr) [reply .github/CODEOWNERS:18]: and fixed.

File: avm/ptn/azd/ml-hub-dependencies/main.bicep (hunk 3 above)
Comment (AlexanderSehr) [reply avm/ptn/azd/ml-hub-dependencies/main.bicep:311]: How come this was resolved? The parameter is still missing?

File: avm/ptn/azd/ml-hub-dependencies/main.bicep [hunk 9]
```diff
@@ -0,0 +1,475 @@
...
+          value: 'For more information, see https://aka.ms/avm/TelemetryInfo'
+        }
+      }
+    }
+  }
+}
+
+module keyVault 'br/public:avm/res/key-vault/vault:0.7.1' = if (enableTelemetry) {
+  name: '${uniqueString(deployment().name, location)}-keyvault'
+  params: {
```
Comment (AlexanderSehr) [inline avm/ptn/azd/ml-hub-dependencies/main.bicep:297]: Missing passthru of `enableTelemetry` parameter

File: avm/ptn/azd/ml-hub-dependencies/main.bicep (hunk 4 above)
Comment (AlexanderSehr) [reply avm/ptn/azd/ml-hub-dependencies/main.bicep:328]: How come this was resolved? The parameter is still missing?

File: avm/ptn/azd/ml-hub-dependencies/main.bicep (hunk 5 above)
Comment (AlexanderSehr) [reply avm/ptn/azd/ml-hub-dependencies/main.bicep:344]: How come this was resolved? The parameter is still missing?

File: avm/ptn/azd/ml-hub-dependencies/main.bicep (hunk 6 above)
Comment (AlexanderSehr) [reply avm/ptn/azd/ml-hub-dependencies/main.bicep:355]: How come this was resolved? The parameter is still missing?

File: avm/ptn/azd/ml-hub-dependencies/main.bicep (hunk 7 above)
Comment (AlexanderSehr) [reply avm/ptn/azd/ml-hub-dependencies/main.bicep:366]: How come this was resolved? The parameter is still missing?

File: avm/ptn/azd/ml-hub-dependencies/main.bicep (hunk 8 above)
Comment (AlexanderSehr) [reply avm/ptn/azd/ml-hub-dependencies/main.bicep:377]: How come this was resolved? The parameter is still missing?

File: avm/ptn/azd/ml-hub-dependencies/main.bicep (hunk 2 above)
Comment (AlexanderSehr) [reply avm/ptn/azd/ml-hub-dependencies/main.bicep]: This should not have ben resolved.
Please note @Menghua1, I'm talking about a passthru - not a condition of the entire deployment.

You'd always want to deployment, independent of the telemetry. What I'm asking for is adding `enableTelemetry: enableTelemetry` to to params block of each module reference. Please do that and remove the condition. And don't resolve the comments, but just reply that you've completed the ask 😉

**Comment by Menghua1** [comment]
@AlexanderSehr Could you please check the [latest commit](https://github.com/Azure/bicep-registry-modules/pull/3241/commits/57090c1d666acdd866db57688deca653d79d0c3c)? The issues mentioned above have been fixed. Thank you!

**Comment by AlexanderSehr** [comment]
> @AlexanderSehr Could you please check the [latest commit](https://github.com/Azure/bicep-registry-modules/pull/3241/commits/57090c1d666acdd866db57688deca653d79d0c3c)? The issues mentioned above have been fixed. Thank you!

oh I did 😉 Please read through my comments - in one of them I pointed out why the 'fix' was unfortunately not the correct one. There should not be any additional conditions added to the module deployments but just a parameter added in the params block

**Comment by Menghua1** [comment]
@AlexanderSehr I apologize for the oversight. Thank you for your feedback! I have submitted a new commit to address the issues correctly. Thank you very much.

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**